llvm/lib/CASObjectFormats/FlatV1.cpp

```diff
@@ -436,28 +424,43 @@ Expected<BlockRef> BlockRef::create(CompileUnitBuilder &CUB,
     Fixups.back().Offset = E->getOffset();
   }
 
-  // Normalize content and put it at the end of the block.
-  // FIXME: assume current block alignment is the alignment for padding and just
-  // pad every block.
-  Optional<Align> TrailingNopsAlignment;
-  TrailingNopsAlignment = Align{Block.getAlignment()};
-  StringRef Content(Block.getContent().begin(), Block.getSize());
-  SmallString<1024> MutableContentStorage;
-  if (Error E = helpers::canonicalizeContent(CUB.TT.getArch(), Content,
-                                             MutableContentStorage, Fixups,
-                                             TrailingNopsAlignment)
-                    .moveInto(Content))
-    return E;
-
+  // Create BlockData.
+  SmallString<1024> EncodeContent;
+  if (!Block.isZeroFill()) {
+    SmallString<1024> MutableContentStorage;
+    // Normalize content and put it at the end of the block.
+    // FIXME: assume current block alignment is the alignment for padding and
+    // just pad every block.
+    Optional<Align> TrailingNopsAlignment;
+    TrailingNopsAlignment = Align{Block.getAlignment()};
+    StringRef Content(Block.getContent().begin(), Block.getSize());
+    if (Error E = helpers::canonicalizeContent(CUB.TT.getArch(), Content,
+                                               MutableContentStorage, Fixups,
+                                               TrailingNopsAlignment)
+                      .moveInto(Content))
+      return E;
+    data::BlockData::encode(Content.size(), Block.getAlignment(),
+                            Block.getAlignmentOffset(), Content,
+                            ArrayRef<Fixup>(), EncodeContent);
+  } else
+    data::BlockData::encode(Block.getSize(), Block.getAlignment(),
+                            Block.getAlignmentOffset(), None,
+                            ArrayRef<Fixup>(), EncodeContent);
+
+  StringRef BlockData(EncodeContent);
   // Encode content first with size and data.
   if (UseBlockContentNode) {
-    if (auto E = CUB.createAndReferenceContent(Content))
+    if (auto E = CUB.createAndReferenceContent(BlockData))
       return E;
   } else {
-    encoding::writeVBR8(Content.size(), B->Data);
-    B->Data.append(Content);
+    encoding::writeVBR8(BlockData.size(), B->Data);
+    B->Data.append(BlockData);
   }
 
+  // For zerofill block, we can just return here because it has no edges.
+  if (Block.isZeroFill())
+    return get(B->build());
```

Comment:
Interesting. `main` branch has this assertion but not on experimental/cas/main:
```
/// Add an edge to this block.
void addEdge(Edge::Kind K, Edge::OffsetT Offset, Symbol &Target,
                  Edge::AddendT Addend) {
       assert(!isZeroFill() && "Adding edge to zero-fill block?");
       Edges.push_back(Edge(K, Offset, Target, Addend));
}
```

Reply:
Aha, I guess if `main` has that, it can't have edges :).

Reply:
I think on this branch, it can have edges. I might have dropped some of those on the floor. Since it is a no-op if zero fill session have no edges, I update the patch just to pretend zerofill can have edges.